libnestutil/numerics.cpp

```diff
@@ -126,3 +131,125 @@ is_integer( double n )
   // factor 4 allows for two bits of rounding error
   return frac_part < 4 * n * std::numeric_limits< double >::epsilon();
 }
+
+long
+mod_inverse( long a, long m )
+{
+  /*
+   The implementation here is based on the extended Euclidean algorithm which
+   solves
+
+     a x + m y = gcd( a, m ) = 1 mod m
+
+   for x and y. Note that the my term is zero mod m, so the equation is equivalent
+   to
+
+     a x = 1 mod m
+
+   Since we only need x, we can ignore y and use just half of the algorithm.
+
+   For details on the algorithm, see D. E. Knuth, The Art of Computer Programming,
+   ch 4.5.2, Algorithm X (vol 2), and ch 1.2.1, Algorithm E (vol 1).
+   */
+
+  assert( 0 <= a );
+
```

Comment:
One could add `assert( 2 <= m );` at this point to make explicit that `m=1` does not make sense (even it will also fail in the last `assert()`-statement. Moreover, `m=0` leads to a floating point exception. Of course, both of these pathological cases cannot occur as long as `mod_inverse()` is called from `first_index()` as it is implemented now.